File: src/main/java/org/hisp/dhis/expression/ast/Nodes.java
package org.hisp.dhis.expression.ast;

import org.hisp.dhis.expression.spi.DataItem;
import org.hisp.dhis.expression.spi.DataItemModifiers;
import org.hisp.dhis.expression.spi.DataItemType;
import org.hisp.dhis.expression.spi.ID;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.function.BiFunction;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;
import java.util.stream.Stream;

import static java.lang.Integer.parseInt;
import static java.lang.String.format;
import static java.util.stream.Collectors.toList;

/**
 * Implements the different {@link Node} types in the AST.
 *
 * @author Jan Bernitt
 */
public interface Nodes {

    abstract class AbstractNode<T> implements Node<T> {

        final NodeType type;
        final String rawValue;
        final T value;

        AbstractNode(NodeType type, String rawValue, Function<String, T> converter) {
            this(type, rawValue, converter, (val, ex) -> ex);
        }

        AbstractNode(NodeType type, String rawValue, Function<String, T> converter, BiFunction<String,RuntimeException,RuntimeException> rethrowAs) {
            this.type = type;
            this.rawValue = rawValue;
            try {
                this.value = converter.apply(rawValue);
            } catch (RuntimeException ex) {
                throw rethrowAs.apply(rawValue,ex);
            }
        }

        static <E extends Enum<E>> BiFunction<String,RuntimeException,RuntimeException> rethrowAs(Class<E> valueType, Function<E,String> toText) {
            return (rawValue, ex) -> new IllegalArgumentException(format("Invalid %s option: '%s'%n\toptions are: %s",
                    valueType.getSimpleName(), rawValue, Stream.of(valueType.getEnumConstants()).map(toText).collect(toList())));
        }

        @Override
        public final NodeType getType() {
            return type;
        }

        @Override
        public final String getRawValue() {
            return rawValue;
        }

        @Override
        public final T getValue()
        {
            return value;
        }

        @Override
        public void visit(Consumer<Node<?>> visitor, Predicate<Node<?>> filter) {
            if (filter.test(this)) {
                visitor.accept(this);
            }
        }

        @Override
        public String toString() {
            StringBuilder str = new StringBuilder();
            toString(str, "");
            return str.toString();
        }

        void toString(StringBuilder str, String indent) {
            str.append(indent).append(getClass().getSimpleName())
                    .append('[').append(getType().name()).append(" ").append(getRawValue()).append("]\n");
        }
    }

    abstract class ComplexNode<T> extends AbstractNode<T> {

        private List<Node<?>> children = new ArrayList<>();

        ComplexNode(NodeType type, String rawValue, Function<String, T> converter) {
            super(type, rawValue, converter);
        }

        ComplexNode(NodeType type, String rawValue, Function<String, T> converter, BiFunction<String, RuntimeException, RuntimeException> rethrowAs) {
            super(type, rawValue, converter, rethrowAs);
        }

        @Override
        public final int size() {
            return children.size();
        }

        @Override
        public final Node<?> child(int index) {
            return children.get(index);
        }

        @Override
        public Stream<Node<?>> children() {
            return children.stream();
        }

        @Override
        public final void addChild(Node<?> child) {
            children.add(child);
        }

        @Override
        public final void transform(BiFunction<Node<?>,List<Node<?>>,List<Node<?>>> transformer) {
            children = transformer.apply(this, children);
            children.forEach(c -> c.transform(transformer));
        }

        @Override
        public final void visit(Consumer<Node<?>> visitor, Predicate<Node<?>> filter) {
            super.visit(visitor, filter);
            children.forEach(child -> child.visit(visitor, filter));
        }

        @Override
        final void toString(StringBuilder str, String indent) {
            super.toString(str, indent);
            for (Node<?> c : children)
                ((AbstractNode<?>)c).toString(str, indent+"  ");
        }
    }

    final class ParenthesesNode extends ComplexNode<String> {

        public ParenthesesNode(NodeType type, String rawValue) {
            super(type, rawValue, Function.identity());
        }

        @Override
        public ValueType getValueType() {
            return child(0).getValueType();
        }
    }

    final class ArgumentNode extends ComplexNode<Integer> {

        public ArgumentNode(NodeType type, String rawValue) {
            super(type, rawValue, Integer::valueOf);
        }

        @Override
        public ValueType getValueType() {
            return size() == 1 ? child(0).getValueType() : ValueType.UNKNOWN;
        }
    }

    final class FunctionNode extends ComplexNode<NamedFunction> {

        public FunctionNode(NodeType type, String rawValue) {
            super(type, rawValue, NamedFunction::fromName, rethrowAs(NamedFunction.class, NamedFunction::getName));
        }

        @Override
        public ValueType getValueType() {
            ValueType type = getValue().getValueType();
            return !type.isSame() ? type : child(getValue().getParameterTypes().indexOf(ValueType.SAME)).getValueType();
        }
    }

    final class ModifierNode extends ComplexNode<DataItemModifier> {

        public ModifierNode(NodeType type, String rawValue) {
            super(type, rawValue, DataItemModifier::valueOf);
        }

        @Override
        public ValueType getValueType() {
            return getValue().getValueType();
        }
    }

    final class DataItemNode extends ComplexNode<DataItemType> {

        /**
         * This list of modifiers is always empty from pure parsing.
         * It is used to aggregate the effective modifiers in AST transformation step.
         */
        private final List<Node<?>> modifiers = new ArrayList<>();

        public DataItemNode(NodeType type, String rawValue) {
            super(type, rawValue, DataItemType::fromSymbol, rethrowAs(DataItemType.class, DataItemType::getSymbol));
        }

        @Override
        public void addModifier(Node<?> mod) {
            this.modifiers.add(mod);
        }

        @Override
        public Iterable<Node<?>> modifiers() {
            return modifiers;
        }

        @Override
        public ValueType getValueType() {
            return ValueType.UNKNOWN;
        }

        @Override
        public DataItem toDataItem() {
            List<List<ID>> idGroups = new ArrayList<>(List.of(List.of(), List.of(), List.of()));
            DataItemType itemType = getValue();
            for (int i = 0; i < size(); i++) {
                Node<?> arg = child(i);
                Node<?> argC0 = arg.child(0);
                ID.Type type = argC0.getType() == NodeType.IDENTIFIER && argC0.getValue() instanceof Tag
                        ? ((Tag)argC0.getValue()).getIdType()
                        : itemType.getType(size(), i);
                idGroups.set(i, arg.children()
                        .filter(n -> n.getType() == NodeType.UID)
                        .map(n -> new ID(type, n.getRawValue()))
                        .collect(toList()));
            }
            return new DataItem(itemType, idGroups.get(0).get(0), idGroups.get(1), idGroups.get(2), dataItemModifiersOf(modifiers));
        }

        static DataItemModifiers dataItemModifiersOf(List<Node<?>> modifiers) {
            DataItemModifiers.DataItemModifiersBuilder mods = DataItemModifiers.builder();
            modifiers.forEach(mod -> {
                Supplier<Object> value = () -> mod.child(0).child(0).getValue();
                switch ((DataItemModifier)mod.getValue()) {
                        case aggregationType: mods.aggregationType( (AggregationType) value.get()); break;
                        case maxDate: mods.maxDate((LocalDate) value.get()); break;
                        case minDate: mods.minDate( (LocalDate) value.get()); break;
                        case periodOffset: mods.periodOffset( (Integer) value.get()); break;
                        case stageOffset: mods.stageOffset((Integer) value.get()); break;
                        case yearToDate: mods.yearToDate( true); break;
                        case periodAggregation: mods.periodAggregation(true); break;
                    }
                });
            return mods.build();
        }
    }

    abstract class SimpleNode<T> extends AbstractNode<T> {

        SimpleNode(NodeType type, String rawValue, Function<String, T> converter) {
            super(type, rawValue, converter);
        }

        SimpleNode(NodeType type, String rawValue, Function<String, T> converter, BiFunction<String, RuntimeException, RuntimeException> rethrowAs) {
            super(type, rawValue, converter, rethrowAs);
        }

        @Override
        public ValueType getValueType() {
            return ValueType.STRING;
        }
    }

    class TextNode extends SimpleNode<String> {

        public TextNode(NodeType type, String rawValue) {
            super(type, rawValue, Function.identity());
        }

    }

    final class VariableNode extends ComplexNode<VariableType> {

        public VariableNode(NodeType type, String rawValue) {
            super(type, rawValue, VariableType::fromSymbol);
        }

        @Override
        public ValueType getValueType() {
            return isEmpty() ? ValueType.UNKNOWN : child(0).getValueType();
        }
    }

    final class Utf8StringNode extends SimpleNode<String> {

        public Utf8StringNode(NodeType type, String rawValue) {
            super(type, rawValue, Utf8StringNode::decode);
        }

        static String decode(String rawValue) {
            if (rawValue.indexOf('\\') < 0) {
                return rawValue; // no special characters
            }
            StringBuilder str = new StringBuilder();
            char[] chars = rawValue.toCharArray();
            for (int i = 0; i < chars.length; i++) {
                char c = chars[i];
                if (c == '\\') {
                    c = chars[++i];
                    if (c == 'u') {
                        str.appendCodePoint(parseInt(new String(new char[] { chars[++i], chars[++i], chars[++i], chars[++i]}), 16));
                    } else if (c >= '0' && c <= '9') {
                        str.appendCodePoint(parseInt(new String(new char[] { chars[++i], chars[++i], chars[++i] }), 8));
                    } else {
                        switch (c) {
                            case 'b': str.append('\b'); break;
                            case 't': str.append('\t'); break;
                            case 'n': str.append('\n'); break;
                            case 'f': str.append('\f'); break;
                            case 'r': str.append('\r'); break;
                            default: str.append(c); // this is the escaped character
                        }
                    }
                } else {
                    str.append(c);
                }
            }
            return str.toString();
        }
    }

    final class UnaryOperatorNode extends ComplexNode<UnaryOperator> {

        public UnaryOperatorNode(NodeType type, String rawValue) {
            super(type, rawValue, UnaryOperator::fromSymbol, rethrowAs(UnaryOperator.class, UnaryOperator::getSymbol));
        }

        @Override
        public ValueType getValueType() {
            ValueType type = getValue().getValueType();
            return type.isSame() ? child(0).getValueType() : type;
        }
    }

    final class BinaryOperatorNode extends ComplexNode<BinaryOperator> {

        public BinaryOperatorNode(NodeType type, String rawValue) {
            super(type, rawValue, BinaryOperator::fromSymbol, rethrowAs(BinaryOperator.class, BinaryOperator::getSymbol));
        }

        @Override
        public ValueType getValueType() {
            return getValue().getValueType();
        }
    }

    final class BooleanNode extends SimpleNode<Boolean> {

        public  BooleanNode(NodeType type, String rawValue) {
            super(type, rawValue, Boolean::valueOf);
        }

        @Override
        public ValueType getValueType() {
            return ValueType.BOOLEAN;
        }
    }

    final class NumberNode extends SimpleNode<Double> {

        public  NumberNode(NodeType type, String rawValue) {
            super(type, rawValue, Double::valueOf);
        }

        @Override
        public ValueType getValueType() {
            return ValueType.NUMBER;
        }
    }

    final class IntegerNode extends SimpleNode<Integer> {

        public IntegerNode(NodeType type, String rawValue) {
            super(type, rawValue, Integer::valueOf);
        }

        @Override
        public ValueType getValueType() {
            return ValueType.NUMBER;
        }
    }

    final class DateNode extends SimpleNode<LocalDate> {

        public DateNode(NodeType type, String rawValue) {
            super(type, rawValue, LocalDate::parse);
        }

        @Override
        public ValueType getValueType() {
            return ValueType.DATE;
        }
    }

    final class ConstantNode extends SimpleNode<Void> {

        public ConstantNode(NodeType type, String rawValue) {
            super(type, rawValue, str -> null);
        }

        @Override
        public ValueType getValueType() {
            return ValueType.SAME;
        }
    }

    final class ReportingRateTypeNode extends SimpleNode<ReportingRateType> {

        public ReportingRateTypeNode(NodeType type, String rawValue) {
            super(type, rawValue, ReportingRateType::valueOf, rethrowAs(ReportingRateType.class, ReportingRateType::name));
        }
    }

    final class ProgramVariableNode extends SimpleNode<ProgramVariable> {

        public ProgramVariableNode(NodeType type, String rawValue) {
            super(type, rawValue, ProgramVariable::valueOf, rethrowAs(ProgramVariable.class, ProgramVariable::name));
        }
    }

    final class NamedValueNode extends SimpleNode<NamedValue>
    {
        public NamedValueNode(NodeType type, String rawValue) {
            super(type, rawValue, NamedValue::valueOf, rethrowAs(NamedValue.class, NamedValue::name));
        }
    }

    final class TagNode extends SimpleNode<Tag> {
        public TagNode(NodeType type, String rawValue) {
            super(type, rawValue, Tag::valueOf, rethrowAs(Tag.class, Tag::name));
        }
    }

    final class AggregationTypeNode extends SimpleNode<AggregationType> {
        public AggregationTypeNode(NodeType type, String rawValue) {
            super(type, rawValue, AggregationType::valueOf, rethrowAs(AggregationType.class, AggregationType::name));
        }
    }
}

File: src/jvmMain/java/org/hisp/dhis/lib/expression/ast/Nodes.kt
package org.hisp.dhis.lib.expression.ast

import org.hisp.dhis.lib.expression.spi.*
import org.hisp.dhis.lib.expression.ast.NodeAnnotations.Whitespace;
import java.time.LocalDate

/**
 * Implements the different [Node] types in the AST.
 *
 * @author Jan Bernitt
 */
object Nodes {

    abstract class AbstractNode<T> @JvmOverloads internal constructor(
        private val type: NodeType,
        private val rawValue: String,
        converter: (String) -> T,
        rethrowAs: (String, RuntimeException) -> RuntimeException? = { _: String, ex: RuntimeException -> ex }
    ) : Node<T> {
        private var value: T
        private var start: Position? = null
        private var end: Position? = null
        private var whitespace: Whitespace = Whitespace.DEFAULT

        init {
            try {
                value = converter(rawValue)
            } catch (ex: RuntimeException) {
                throw rethrowAs(rawValue, ex)!!
            }
        }

        override fun setStart(start: Position?) {
            this.start = start
        }

        override fun setEnd(end: Position?) {
            this.end = end
        }

        override fun setWhitespace(whitespace: Whitespace) {
            this.whitespace = whitespace
        }

        override fun getWhitespace(): Whitespace {
            return whitespace
        }

        override fun getStart(): Position? {
            return start
        }

        override fun getEnd(): Position? {
            return end
        }

        override fun getType(): NodeType {
            return type
        }

        override fun getRawValue(): String {
            return rawValue
        }

        override fun getValue(): T {
            return value
        }

        override fun visit(visitor: (Node<*>) -> Unit, filter: (Node<*>) -> Boolean) {
            if (filter(this)) {
                visitor(this)
            }
        }

        override fun toString(): String {
            val str = StringBuilder()
            toString(str, "")
            return str.toString()
        }

        open fun toString(str: StringBuilder, indent: String) {
            str.append(indent).append(javaClass.simpleName)
                .append('[').append(getType().name).append(" ").append(getRawValue()).append("]\n")
        }

        companion object {
            fun <E : Enum<E>?> rethrowAs(
                name: String?,
                values: Collection<E>,
                toText: (E) -> String
            ): (String, RuntimeException) -> RuntimeException {
                return { rawValue: String, _: RuntimeException ->
                    IllegalArgumentException(
                        String.format(
                            "Invalid %s option: '%s'%n\toptions are: %s",
                            name,
                            rawValue,
                            values.map(toText)
                        )
                    )
                }
            }
        }
    }

    abstract class ComplexNode<T> : AbstractNode<T> {

        private var children: MutableList<Node<*>> = ArrayList()

        internal constructor(type: NodeType, rawValue: String, converter: (String) -> T)
                : super(type, rawValue, converter)

        internal constructor(
            type: NodeType,
            rawValue: String,
            converter: (String) -> T,
            rethrowAs: (String, RuntimeException) -> RuntimeException
        ) : super(type, rawValue, converter, rethrowAs)

        override fun size(): Int {
            return children.size
        }

        override fun child(index: Int): Node<*> {
            return children[index]
        }

        override fun children(): Sequence<Node<*>> {
            return children.asSequence()
        }

        override fun addChild(child: Node<*>): Node<T> {
            children.add(child)
            return this
        }

        override fun transform(transformer: (Node<*>, List<Node<*>>) -> List<Node<*>>) {
            children = transformer(this, children).toMutableList()
            children.forEach { c: Node<*> -> c.transform(transformer) }
        }

        override fun visit(visitor: (Node<*>) -> Unit, filter: (Node<*>) -> Boolean) {
            super.visit(visitor, filter)
            children.forEach { child: Node<*> -> child.visit(visitor, filter) }
        }

        override fun toString(str: StringBuilder, indent: String) {
            super.toString(str, indent)
            for (c in children) (c as AbstractNode<*>).toString(str, "$indent  ")
        }
    }

    class ParenthesesNode(type: NodeType, rawValue: String) :
            ComplexNode<String>(type, rawValue, { it }) {
        override fun getValueType(): ValueType {
            return child(0).getValueType()
        }
    }

    class ArgumentNode(type: NodeType, rawValue: String) :
            ComplexNode<Int>(type, rawValue, Integer::valueOf) {
        override fun getValueType(): ValueType {
            return if (size() == 1) child(0).getValueType() else ValueType.MIXED
        }
    }

    class FunctionNode(type: NodeType, rawValue: String) :
            ComplexNode<NamedFunction>(
                type, rawValue, NamedFunction::fromName, rethrowAs(NamedFunction::class.simpleName,
                    NamedFunction.entries, NamedFunction::getName)) {
        override fun getValueType(): ValueType {
            val type = getValue().getValueType()
            return if (!type.isSame()) type else child(getValue().parameterTypes.indexOf(ValueType.SAME)).getValueType()
        }
    }

    class ModifierNode(type: NodeType, rawValue: String) :
            ComplexNode<DataItemModifier>(type, rawValue, DataItemModifier::valueOf) {
        override fun getValueType(): ValueType {
            return getValue().getValueType()
        }
    }

    abstract class ModifiedNode<T> : ComplexNode<T> {
        /**
         * This list of modifiers is always empty from pure parsing. It is used to aggregate the effective modifiers in
         * AST transformation step.
         */
        private val modifiers: MutableList<Node<*>> = ArrayList()

        internal constructor(type: NodeType, rawValue: String, converter: (String) -> T)
                : super(type, rawValue, converter)

        internal constructor(
            type: NodeType,
            rawValue: String,
            converter: (String) -> T,
            rethrowAs: (String, RuntimeException) -> RuntimeException
        ) : super(type, rawValue, converter, rethrowAs)

        override fun addModifier(modifier: Node<*>): Node<T> {
            modifiers.add(modifier)
            return this
        }

        override fun modifiers(): Iterable<Node<*>> {
            return modifiers
        }

        val queryModifiers: QueryModifiers
            get() {
                var mods = QueryModifiers()
                val sum = { a: Int?, b: Int -> if (a == null) b else a + b }
                for (mod in modifiers) {
                    val value = { mod.child(0).child(0).getValue() }
                    mods = when (mod.getValue() as DataItemModifier) {
                        DataItemModifier.aggregationType ->
                            mods.copy(aggregationType = value() as AggregationType)
                        DataItemModifier.maxDate ->
                            mods.copy(maxDate = value() as LocalDate)
                        DataItemModifier.minDate ->
                            mods.copy(minDate = value() as LocalDate)
                        DataItemModifier.periodOffset ->
                            mods.copy(periodOffset = sum(mods.periodOffset, value() as Int))
                        DataItemModifier.stageOffset ->
                            mods.copy(stageOffset = sum(mods.stageOffset, value() as Int))
                        DataItemModifier.yearToDate ->
                            mods.copy(yearToDate = true)
                        DataItemModifier.periodAggregation ->
                            mods.copy(periodAggregation = true)
                        DataItemModifier.subExpression ->
                            mods.copy(subExpression = value()as String)
                    }
                }
                return mods
            }
    }

    class DataItemNode(type: NodeType, rawValue: String) :
            ModifiedNode<DataItemType>(type, rawValue, DataItemType::fromSymbol, rethrowAs(
                DataItemType::class.simpleName, DataItemType.entries, DataItemType::getSymbol
            )) {
        /**
         * Just to cache the result
         */
        private var dataItemValue: DataItem? = null

        override fun getValueType(): ValueType {
            return ValueType.MIXED
        }

        override fun toDataItem(): DataItem {
            if (dataItemValue != null) return dataItemValue as DataItem
            val idGroups: MutableList<List<ID>> = ArrayList(listOf(listOf(), listOf(), listOf()))
            val itemType = getValue()
            for (i in 0 until size()) {
                val arg = child(i)
                val argC0 = arg.child(0)
                val type = if (argC0.getType() === NodeType.IDENTIFIER && argC0.getValue() is Tag) (argC0.getValue() as Tag).idType
                else itemType.getType(
                    size(),
                    i
                )
                var ids = arg.children()
                    .filter { n: Node<*> -> n.getType() === NodeType.UID }
                    .map { n: Node<*> -> ID(type, n.getRawValue()) }
                    .toList()
                if (ids.isEmpty() && arg.size() == 1 && arg.child(0).getType() === NodeType.IDENTIFIER) {
                    ids = listOf(ID(type, arg.child(0).getRawValue()))
                }
                idGroups[i] = ids
            }
            dataItemValue = DataItem(itemType, idGroups[0][0], idGroups[1], idGroups[2], queryModifiers)
            return dataItemValue!!
        }
    }

    abstract class SimpleNode<T> : AbstractNode<T> {
        internal constructor(type: NodeType, rawValue: String, converter: (String) -> T)
                : super(type, rawValue, converter)

        internal constructor(
            type: NodeType,
            rawValue: String,
            converter: (String) -> T,
            rethrowAs: (String, RuntimeException) -> RuntimeException
        ) : super(type, rawValue, converter, rethrowAs)

        override fun getValueType(): ValueType {
            return ValueType.STRING
        }
    }

    class TextNode(type: NodeType, rawValue: String) : SimpleNode<String>(type, rawValue, { it })

    class VariableNode(type: NodeType, rawValue: String) :
            ModifiedNode<VariableType>(type, rawValue, VariableType::fromSymbol) {
        /**
         * Might be provided after AST has been constructed to specify more precisely what type a variable has
         */
        private var actualValueType: ValueType? = null

        override fun toVariable(): Variable? {
            return if (getValue() !== VariableType.PROGRAM) {
                null
            }
            else Variable(child(0).getValue() as ProgramVariable, queryModifiers)
        }

        fun setActualValueType(actualValueType: ValueType?) {
            this.actualValueType = actualValueType
        }

        override fun getValueType(): ValueType {
            // the type of variable value is unknown statically
            return if (actualValueType == null) ValueType.MIXED else actualValueType!!
        }
    }

    class Utf8StringNode(type: NodeType, rawValue: String) :
            SimpleNode<String>(type, rawValue, { raw: String -> decode(raw) }) {
        companion object {
            fun decode(rawValue: String): String {
                if (rawValue.indexOf('\\') < 0) {
                    return rawValue // no special characters
                }
                val str = StringBuilder()
                val chars = rawValue.toCharArray()
                var i = 0
                while (i < chars.size) {
                    var c = chars[i]
                    if (c == '\\') {
                        c = chars[++i]
                        when (c) {
                            'u' -> {
                                str.appendCodePoint(
                                    String(charArrayOf(chars[++i], chars[++i], chars[++i], chars[++i])).toInt(16)
                                )
                            }
                            in '0'..'9' -> {
                                str.appendCodePoint(String(charArrayOf(chars[++i], chars[++i], chars[++i])).toInt(8))
                            }
                            else -> {
                                when (c) {
                                    'b' -> str.append('\b')
                                    't' -> str.append('\t')
                                    'n' -> str.append('\n')
                                    'f' -> str.append('\u000c')
                                    'r' -> str.append('\r')
                                    else -> str.append(c) // this is the escaped character
                                }
                            }
                        }
                    }
                    else {
                        str.append(c)
                    }
                    i++
                }
                return str.toString()
            }
        }
    }

    class UnaryOperatorNode(type: NodeType, rawValue: String) :
            ComplexNode<UnaryOperator>(type, rawValue, UnaryOperator::fromSymbol,
                rethrowAs(UnaryOperator::class.simpleName, UnaryOperator.entries, UnaryOperator::symbol)) {

                override fun getValueType(): ValueType {
            val type = getValue().getValueType()
            return if (type.isSame()) child(0).getValueType() else type
        }
    }

    class BinaryOperatorNode(type: NodeType, rawValue: String) : ComplexNode<BinaryOperator>(
        type, rawValue, BinaryOperator::fromSymbol,
        rethrowAs(BinaryOperator::class.simpleName, BinaryOperator.entries, BinaryOperator::symbol))
    {
        override fun getValueType(): ValueType {
            return getValue().getValueType()
        }
    }

    class BooleanNode(type: NodeType, rawValue: String) :
            SimpleNode<Boolean>(type, rawValue, String::toBoolean) {

        override fun getValueType(): ValueType {
            return ValueType.BOOLEAN
        }
    }

    class NumberNode(type: NodeType, rawValue: String) :
            SimpleNode<Double>(type, rawValue, String::toDouble) {

        override fun getValueType(): ValueType {
            return ValueType.NUMBER
        }
    }

    class IntegerNode(type: NodeType, rawValue: String) :
            SimpleNode<Int>(type, rawValue, String::toInt) {

        override fun getValueType(): ValueType {
            return ValueType.NUMBER
        }
    }

    class DateNode(type: NodeType, rawValue: String) :
            SimpleNode<LocalDate>(type, rawValue, LocalDate::parse) {

        override fun getValueType(): ValueType {
            return ValueType.DATE
        }
    }

    class ConstantNode(type: NodeType, rawValue: String) :
            SimpleNode<Unit>(type, rawValue, { _: String -> null }) {

        override fun getValueType(): ValueType {
            return ValueType.SAME
        }
    }

    class ReportingRateTypeNode(type: NodeType, rawValue: String) :
            SimpleNode<ReportingRateType>(
                type, rawValue, ReportingRateType::valueOf, rethrowAs(
                    ReportingRateType::class.simpleName, ReportingRateType.entries, ReportingRateType::name))

    class ProgramVariableNode(type: NodeType, rawValue: String) :
            SimpleNode<ProgramVariable>(type, rawValue, ProgramVariable::valueOf, rethrowAs(
                ProgramVariable::class.simpleName, ProgramVariable.entries, ProgramVariable::name))

    class NamedValueNode(type: NodeType, rawValue: String) :
            SimpleNode<NamedValue>(type, rawValue, NamedValue::valueOf, rethrowAs(
                NamedValue::class.simpleName, NamedValue.entries, NamedValue::name))

    class TagNode(type: NodeType, rawValue: String) : SimpleNode<Tag>(
        type, rawValue, Tag::valueOf, rethrowAs(Tag::class.simpleName, Tag.entries, Tag::name))

    class AggregationTypeNode(type: NodeType, rawValue: String) :
            SimpleNode<AggregationType>(type, rawValue, AggregationType::valueOf,
            rethrowAs(AggregationType::class.simpleName, AggregationType.entries, AggregationType::name))

    fun supplySubExpressionSQL(root: Node<*>) {
        root.transform { node: Node<*>, children: List<Node<*>> ->
            if (node.getValue() === NamedFunction.subExpression) {
                children.forEach { child: Node<*> ->
                    child.visit(NodeType.DATA_ITEM) { modified: Node<*> ->
                        modified
                            .addModifier(
                                ModifierNode(NodeType.MODIFIER, DataItemModifier.subExpression.name)
                                    .addChild(
                                        ArgumentNode(NodeType.ARGUMENT, "0")
                                            .addChild(
                                                TextNode(
                                                    NodeType.STRING,
                                                    "@" + System.identityHashCode(node)
                                                )
                                            )
                                    )
                            )
                    }
                }
            }
            children
        }
    }

    /**
     * Modifiers affect data items only. However, they can be applied to data items directly or indirectly. Within a
     * function or round bracket that has a modifier all data items within the bracket body are affected.
     *
     *
     * This transformation moves modifiers from being [Node.children] to be added as
     * [Node.addModifier].
     *
     *
     * This transformation should only be applied when the expression should be evaluated including resolving data items
     * to their actual value.
     *
     * @param root the node to start the transformation from.
     */
    fun propagateModifiers(root: Node<*>) {
        root.transform { node: Node<*>, children: List<Node<*>> ->
            val isModifier = { child: Node<*> -> child.getType() === NodeType.MODIFIER }
            if (node.getValue() is NamedFunction && (node.getValue() as NamedFunction).isAggregating()) {
                children.forEach { child: Node<*> ->
                    child.visit(NodeType.DATA_ITEM) { modified: Node<*> ->
                        modified.addModifier(
                            ModifierNode(NodeType.MODIFIER, DataItemModifier.periodAggregation.name)
                        )
                    }
                }
            }
            if (children.none(isModifier)) {
                return@transform children
            }
            // attach any modifier found on this level to any data item in the subtree of the child before them
            for (i in 1 until children.size) {
                val maybeModifier = children[i]
                if (maybeModifier.getType() === NodeType.MODIFIER) {
                    // go back 1 (or more if node before is a modifier)
                    var target = i - 1
                    while (target >= 0 && children[target].getType() === NodeType.MODIFIER) target--
                    if (target >= 0) {
                        val addModifier : (Node<*>) -> Unit = { modified: Node<*> -> modified.addModifier(maybeModifier) }
                        children[target].visit(NodeType.DATA_ITEM, addModifier)
                        children[target].visit(
                            addModifier
                        ) { n: Node<*> -> n.getType() === NodeType.VARIABLE && n.getValue() === VariableType.PROGRAM }
                    }
                }
            }
            children.filterNot(isModifier)
        }
    }
}
